File: backend/src/main/java/com/fksoft/erp/domain/commission/service/data/CommissionStatement.java
package com.fksoft.erp.domain.commission.service.data;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

/**
 * Simple commission statement for one beneficiary over an (optional) period: the commission entries plus the per-status
 * totals (expected / eligible / approved / paid). Informational only — it approves/pays nothing and carries commission
 * + commercial-origin data only (never payroll, tax or accounting data). Visibility is enforced by the service (a
 * seller/representative sees only their own statement; a manager/finance can see any beneficiary).
 *
 * @param beneficiaryId the beneficiary user id
 * @param beneficiaryName the resolved beneficiary name, or {@code null}
 * @param periodFrom the inclusive period start (by creation date), or {@code null} (all time)
 * @param periodTo the inclusive period end (by creation date), or {@code null} (all time)
 * @param entries the commission entries (newest first)
 * @param totals the per-status totals
 */
public record CommissionStatement(
        java.util.UUID beneficiaryId,
        String beneficiaryName,
        LocalDate periodFrom,
        LocalDate periodTo,
        List<CommissionListItem> entries,
        Totals totals) {

    /**
     * The per-status totals over the statement's entries — amount sums and counts for the non-voided lifecycle.
     *
     * @param totalExpected the sum of EXPECTED amounts
     * @param totalEligible the sum of ELIGIBLE amounts
     * @param totalApproved the sum of APPROVED amounts
     * @param totalPaid the sum of PAID amounts
     * @param countExpected the number of EXPECTED entries
     * @param countEligible the number of ELIGIBLE entries
     * @param countApproved the number of APPROVED entries
     * @param countPaid the number of PAID entries
     */
    public record Totals(
            BigDecimal totalExpected,
            BigDecimal totalEligible,
            BigDecimal totalApproved,
            BigDecimal totalPaid,
            long countExpected,
            long countEligible,
            long countApproved,
            long countPaid) {}

    /**
     * Builds a statement, computing the per-status totals from the entries in memory (a per-beneficiary statement is a
     * small result set).
     *
     * @param beneficiaryId the beneficiary user id
     * @param beneficiaryName the resolved beneficiary name, or {@code null}
     * @param from the inclusive period start, or {@code null}
     * @param to the inclusive period end, or {@code null}
     * @param entries the commission entries
     * @return the statement with computed totals
     */
    public static CommissionStatement of(
            java.util.UUID beneficiaryId,
            String beneficiaryName,
            LocalDate from,
            LocalDate to,
            List<CommissionListItem> entries) {
        Totals totals = new Totals(
                sum(entries, "EXPECTED"),
                sum(entries, "ELIGIBLE"),
                sum(entries, "APPROVED"),
                sum(entries, "PAID"),
                count(entries, "EXPECTED"),
                count(entries, "ELIGIBLE"),
                count(entries, "APPROVED"),
                count(entries, "PAID"));
        return new CommissionStatement(beneficiaryId, beneficiaryName, from, to, entries, totals);
    }

    private static BigDecimal sum(List<CommissionListItem> entries, String status) {
        return entries.stream()
                .filter(e -> status.equals(e.status()))
                .map(CommissionListItem::amount)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    private static long count(List<CommissionListItem> entries, String status) {
        return entries.stream().filter(e -> status.equals(e.status())).count();
    }
}

File: frontend/src/app/features/commission/commission-statement/commission-statement.ts
import { Component, OnInit, computed, inject, signal } from '@angular/core';
import { CurrencyPipe, DatePipe } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { RouterLink } from '@angular/router';
import { HttpErrorResponse } from '@angular/common/http';
import { CardModule } from 'primeng/card';
import { TableModule } from 'primeng/table';
import { TagModule } from 'primeng/tag';
import { SelectModule } from 'primeng/select';
import { DatePickerModule } from 'primeng/datepicker';
import { ToggleSwitchModule } from 'primeng/toggleswitch';
import { ButtonModule } from 'primeng/button';
import { MessageModule } from 'primeng/message';
import { AuthService } from '../../../core/auth/auth.service';
import { Responsible } from '../../../core/api/lead.service';
import {
  CommissionListItem,
  CommissionService,
  CommissionStatement,
  CommissionStatus,
} from '../../../core/api/commission.service';

type TagSeverity = 'success' | 'info' | 'warn' | 'secondary' | 'contrast' | 'danger';

const STATUS_LABELS: Record<CommissionStatus, string> = {
  EXPECTED: 'Prevista',
  ELIGIBLE: 'Pendente de aprovação',
  APPROVED: 'Aprovada',
  REJECTED: 'Rejeitada',
  PAID: 'Paga',
  CANCELLED: 'Cancelada',
};

const STATUS_SEVERITY: Record<CommissionStatus, TagSeverity> = {
  EXPECTED: 'info',
  ELIGIBLE: 'warn',
  APPROVED: 'success',
  REJECTED: 'danger',
  PAID: 'success',
  CANCELLED: 'secondary',
};

/**
 * Simple commission statement by beneficiary (Commission Management): an informational read view grouping a
 * beneficiary's commission entries over an optional period with per-status totals (expected / eligible / approved /
 * paid). It respects visibility — a seller/representative sees only their own statement (the beneficiary picker is
 * locked to them); a manager/finance can pick any beneficiary. It approves/pays nothing and shows commission +
 * commercial-origin data only — never payroll, tax or accounting data.
 */
@Component({
  selector: 'app-commission-statement',
  imports: [
    CurrencyPipe,
    DatePipe,
    FormsModule,
    RouterLink,
    CardModule,
    TableModule,
    TagModule,
    SelectModule,
    DatePickerModule,
    ToggleSwitchModule,
    ButtonModule,
    MessageModule,
  ],
  templateUrl: './commission-statement.html',
  styleUrl: './commission-statement.css',
})
export class CommissionStatementPage implements OnInit {
  private readonly commissions = inject(CommissionService);
  protected readonly auth = inject(AuthService);

  protected readonly statement = signal<CommissionStatement | null>(null);
  protected readonly loading = signal(false);
  protected readonly error = signal<string | null>(null);
  protected readonly responsibleOptions = signal<Responsible[]>([]);

  /** Managers can choose any beneficiary; own-tier users are locked to themselves. */
  protected readonly canChooseBeneficiary = computed(() => this.auth.canSeeAllCommissions());

  protected beneficiary: string | null = null;
  protected from: Date | null = null;
  protected to: Date | null = null;
  protected includeVoided = false;

  ngOnInit(): void {
    this.beneficiary = this.auth.userId();
    if (this.canChooseBeneficiary()) {
      this.commissions.responsibles().subscribe({
        next: (list) => this.responsibleOptions.set(list),
      });
    }
    this.load();
  }

  protected statusLabel(status: CommissionStatus): string {
    return STATUS_LABELS[status];
  }

  protected statusSeverity(status: CommissionStatus): TagSeverity {
    return STATUS_SEVERITY[status];
  }

  /** The human-friendly source order code (PC-000n). */
  protected orderCode(n: number): string {
    return 'PC-' + String(n).padStart(4, '0');
  }

  /** Reloads the statement for the chosen beneficiary + period. */
  protected apply(): void {
    this.load();
  }

  private load(): void {
    if (!this.beneficiary) {
      return;
    }
    this.loading.set(true);
    this.error.set(null);
    this.commissions
      .statement(this.beneficiary, toIsoDate(this.from), toIsoDate(this.to), this.includeVoided)
      .subscribe({
        next: (s) => {
          this.statement.set(s);
          this.loading.set(false);
        },
        error: (err: HttpErrorResponse) => {
          this.loading.set(false);
          this.statement.set(null);
          this.error.set(
            err.status === 403
              ? 'Você só pode ver o seu próprio extrato.'
              : 'Não foi possível carregar o extrato de comissões.',
          );
        },
      });
  }

  protected trackEntry(_: number, entry: CommissionListItem): string {
    return entry.id;
  }
}

function toIsoDate(date: Date | null): string | null {
  if (!date) {
    return null;
  }
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
}
